lib/utils.pm

```diff
@@ -1346,6 +1346,8 @@ sub reconnect_mgmt_console {
             select_console 'novalink-ssh', await_console => 0;
         } elsif (check_var('BACKEND', 'pvm_hmc')) {
             select_console 'powerhmc-ssh', await_console => 0;
+            sleep 100;
```

Comment:
do not using sleep , i suppose Oliver will give same comments on this....

Reply:
I checked other code also use sleep. I checked openqa API cannot find other to use.

Reply:
So what exactly is the reason for `sleep` here? Is there a way to sync on some other event, e.g. availability of ssh, etc.? Could you try to increase timeout for the `reconnect_mgmt_console` below? For example:
 `reconnect_mgmt_console(timeout => 500);`. In this case function will wait till it's able to reconnect to the VM and it worked for us just fine. Looks like we won't need any sleep here after that.

Reply:
This sleep to wait system reboot and show grub page. 
For reconnect_mgmt_console(timeout => 500) this timeout only be used for s390x console. It is not used at pvm code. I  have changed this timeout to 1000 before, it cannot wait until grub page show up. Please check https://openqa.nue.suse.com/tests/4338941 at log, it didn't wait 1000s.
I don't know why your case can reconnect successfully.
I used to debug to wait 30s and then wait 60s, please check https://openqa.nue.suse.com/tests/4339637#step/yast2_migration/18 
You can get what did it do during sleep 100s.

Reply:
Still the problem is that you are using `sleep()` or are simulating a sleep-like functionality. This is introducing the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. You *can* use sleep but there are only very limited cases where is the best solution (e.g. yield threads in low-level system code). Here you have the classical problem in software design of [Synchronization](https://en.wikipedia.org/wiki/Synchronization_(computer_science)) . sleep should not be used in place of proper synchronization methods. https://blogs.msmvps.com/peterritchie/2007/04/26/thread-sleep-is-a-sign-of-a-poorly-designed-program/ is one of multiple references that explains this: "The original problem is likely a timing/synchronization issue, ignoring it by hiding it with Thread.Sleep is only going to delay the problem and make it occur in random, hard to reproduce ways." And if the problem does not occur at least time is wasted.

We are waiting because some condition changes some time ... keyword(s) is/are *some time*. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time. For openQA tests in general just try to do what you as a human user would also do. You would not look on your watch waiting for exactly 100s before you look at the screen again, right? :)

In general I think the migration test cases could benefit from using more common code, e.g. test modules like reconnect_mgmt_console , when trying to enable the tests against new backends as most problems have been solved already by other tests, same for ipmi, hyperv, s390x, etc.

Reply:
So let me explain why we are so picky about such sleeps. As of now we have 60+ jobs on powerVM (only 20 use hmc backend, but we novalink ones might be moved there too), and in the worst case scenario it means that we will waste 100 minutes for each build. And this number will get doubled or tripled once more kvm test suites are migrated to powerVM. So in case we cannot find better way, we should at least limit it to the cases where we need it and don't apply it for every single scenario. Hope that helps.

Reply:
Hi, I deleted sleep code. Add wait_grub at the beginning of the wait_boot. So after select_console powerhmc-ssh,it will wait at $self->wait_grub(bootloader_time => $bootloader_time) if (is_pvm); Do you thinks this method is OK for you?
Verified: https://10.160.0.207/tests/4371447